### ISSUE TITLE
Consider adding assertWithMessage(Object) overload

### ISSUE THREAD

**Issue by carl-mastrangelo** [google/truth#498]
When calling into the sub-fields of a composite  object, object the subject is wrong.  Consider:

```java
class Status {
  String message;
  Code code;
}
```

Calling Truth with `assertThat` is not ideal, because the `Status` subject is lost:

```java
assertThat(status.code).isEqualTo(Code.INVALID_ARGUEMNT);
``` 

The code may be different, but to debug the test failure, we need to know what the message was.  Similarly, checking the message may lose what the code originally was.

Making a custom `Subject.Factory` and `Subject` aren't really appropriate here.  They could potentially give better error messages, and more appropriate asserts, but using them makes code very verbose, and precludes reuse.  For example, a custom StatusSubject can't reuse the methods available on the StringSubject class.

`assertWithMessage` seems like the right approach, but it takes Object rather than String, which runs into the verbosity problem again.  Calling assertWithMessage(status.toString()) is really long, and annoying to do if making multiple assertions.  To ameliorate the issue, I propose making an assertWithMessage overload, which takes an Object.   This would make it so the correct subject can be printed when an exception occurs.

**Comment by cpovirk** [comment]
Thanks. We should at least do some research and think about this for a 1.0 release someday.

**Comment by cpovirk** [comment]
Based on some naive searches, I'm estimating that ~3.7% of one-arg `assertWithMessage`/`withMessage` calls take an argument of the form `someExpression.toString()` or `String.valueOf(someExpression)`. (That number is probably a slight overestimate, thanks to the naivety of the search, but at lowest it's 3%.) I'm also seeing some calls to the multi-arg `assertWithMessage`/`withMessage` of the form `("%s", expression)`, which, if we treat them as one-arg calls, raise the number above 3.8%.

That's not a ton, but:

- It's a decent number in absolute terms, even if it's small in relative terms.
- Accepting `Object` seems harmless. (While we regret a similar decision in `checkNotNull(T, Object)`, that's only because it's possible to get the parameters backward, which isn't an issue here.)
- Probably more people would do it if it were easier (as you suggest).

**Comment by cpovirk** [comment]
There are surely other near-users of this. For example, I see a handful of callers who pass `Integer.toString(something)` or `Long.toString(something)`. (I also see `Arrays.toString(something)`, which would be a potential user if we implemented https://github.com/google/truth/issues/584, which I filed after seeing those callers :))

We could even go so far as to speculate that messages of the form `"something " + i` would be just as good including only `i`. Sometimes that's not true, especially with, say, an `isTrue()` assertion. But other times, the message is, besides adding the index, just restating that, say, an equality check is happening. To be fair, though, a bare index number might be easy to miss in the failure message, even though it's on its own line -- or easy to misunderstand.

Still, adding all the above in, we're now over 4.4%. The `"something " + i` search is surely overestimating, but I'm probably missing other good candidates. And I'm not sure the difference between 3% and 4.5% is going to be decisive, anyway :)

**Comment by cpovirk** [comment]
I wonder if anyone would see it as confusing that the parameter to `assertWithMessage` isn't a true "message." I don't think we've heard of any confusion with `Preconditions`, for what it's worth.

Additionally, the fact that a value isn't a `String` doesn't mean that it isn't a "message." I just ran into some code that was building a message with a `StringBuilder` and then passing `message.toString()` to `fail(...)`. For unrelated reasons, I'm changing that code to use `assertWithMessage(...).fail()`, which means that it would now be able to omit the `toString()` call if we implemented this. (I searched and found a handful of other occurrences of a literal `message.toString()` as the argument. Most are `StringBuilder`, and most of the rest are protobufs.)

**Comment by dimo414** [comment]
> I wonder if anyone would see it as confusing that the parameter to assertWithMessage isn't a true "message."

I was going to call this out; `assertWithMessage(myDataStructure)` is pretty unclear, and requiring a string at least prompts you to consider adding more prose (which is the point of "with message"). I don't think it's a huge issue to add, and agree it makes it a little easier to work with, but in theory the slight additional friction leads to better error messages, which maybe is worth the cost.

**Comment by carl-mastrangelo** [comment]
> I was going to call this out; assertWithMessage(myDataStructure) is pretty unclear, and requiring a string at least prompts you to consider adding more prose (which is the point of "with message").

I agree, though my preferred method name, `assertAbout` is already in use :/

**Comment by dimo414** [comment]
Typically the suggestion would be to have a `Subject` if you're intending to "assert about" it. `assertWithMessage()` is (intended to be) more for adding context or details that aren't necessarily part of the assertion itself. Including structured data in that message may make sense, but should it really be the *whole* message?

`assertWithMessage("For request %s", myDataStructure)` is going to provide a more informative error message than a context-less `assertWithMessage(myDataStructure)`. Even `assertWithMessage("%s", myDataStructure)` at least documents "there is no additional context needed" for the message.

**Comment by cpovirk** [comment]
Status update for anyone else following along: All of us so far have fairly weak opinions on this: We see that it's convenient (and may save users from annoying compile errors), but we wonder if it will encourage people to supply overly short "messages" like `5` instead of `at index 5`.

I'm unsure which direction we'll go, but we're thinking about it.

**Comment by cpovirk** [comment]
I'm going to leave this out of 1.0. We can still add an overload someday if we want; we just won't change the existing one in place.
